File: 2-1-remove-dups.py
# 2.1 Remove Dups: Write code to remove duplicates from an unsorted linked list.

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

linked_list = Node(2)

# Add a node
temp = linked_list
while temp.next:
    temp = temp.next
temp.next = Node(3)

# Add a node
temp = linked_list
while temp.next:
    temp = temp.next
temp.next = Node(3)

# Add a node
temp = linked_list
while temp.next:
    temp = temp.next
temp.next = Node(3)

# Add a node 
temp = linked_list
while temp.next:
    temp = temp.next
temp.next = Node(4)

# Print the list
temp = linked_list
while temp:
    print(temp.data)
    temp = temp.next

print()

# Remove duplicates
seen = set()
current_node = linked_list
previous_node = linked_list
while current_node:
    # If the current data has not been seen, add it to the set, set previous to current, then increment current_node
    if current_node.data not in seen:
        seen.add(current_node.data)
        previous_node = current_node
        current_node = current_node.next

    # If the current data has been seen, set previous.next to current_node.next, then increment current_node
    else:
        previous_node.next = current_node.next
        current_node = current_node.next

# Print the list
temp = linked_list
while temp:
    print(temp.data)
    temp = temp.next

File: 2-1-remove-dups.js
// 2.1 Remove Dups: Write code to remove duplicates from an unsorted linked list.

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

let linked_list = new Node(2);

// Prints a linked list
function print_list(linked_list)
{
    let temp = linked_list;
    while (temp)
    {
        console.log(temp.data);
        temp = temp.next;
    }
}

// Print the list as a sanity check
// print_list(linked_list);

// Adds a node to the end of the list
function add_node(linked_list, node)
{
    let current_node = linked_list;
    while (current_node.next)
    {
        current_node = current_node.next;
    }
    current_node.next = node;
}

// Add three nodes to the end of the list
add_node(linked_list, new Node(4));
add_node(linked_list, new Node(6));
add_node(linked_list, new Node(6));
add_node(linked_list, new Node(6));
add_node(linked_list, new Node(8));
add_node(linked_list, new Node(8));

// Print the list as a sanity check
// print_list(linked_list);
// console.log();

// Remove duplicates from an unsorted linked list
function remove_duplicates(linked_list)
{
    let temp = linked_list;
    let seen = new Set();
    while (temp.next)
    {
        seen.add(temp.data);
        if (seen.has(temp.next.data))
        {
            temp.next = temp.next.next;
        }
        else
        {
            temp = temp.next;
        }
    }
}

remove_duplicates(linked_list);
print_list(linked_list);